Support multi-class classification

mshtelma opened the pull request:
Signed-off-by: Michael Shtelma <mshtelma@gmail.com>

This PR introduces support for Multi-Class Classification for MLflow Recipes.
It adds the following functionalities: 
1) Multi-class datasets (datasets containing more than two unique values in the target column) are split on per class basis, so each class is split independently. 
2) Rebalancing of the training dataset is supported only for binary classification problems
3) Added multi-class specific metrics
4) Default metrics are now picked on use case basis (regression, binary classification, and multi-class classification)

## Related Issues/PRs


<!-- Resolve --> #xxx

## What changes are proposed in this pull request?

(Please fill in changes proposed in this fix)

## How is this patch tested?

<!--
If you're unsure about what to test, where to add tests, or how to run tests, please feel free to ask.
-->

- [ x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

<!--
Please describe how you confirmed the proposed feature/bug-fix/change works here. For example, if you fixed an MLflow client API, you could attach the code that didn't work prior to the fix but works now, or if you added a new feature on MLflow UI, you could attach a video that demonstrates the feature.
-->

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

mshtelma committed:
Support Multi Class Classification

Signed-off-by: mshtelma <mshtelma@gmail.com>

mlflow-automation commented:
<!-- documentation preview -->

Documentation preview for df6deb6a188de73dafae34e0a8d99557a8248ffd will be available [here](https://output.circle-artifacts.com/output/job/57d6e44b-d629-4ec8-81af-c8f959332da9/artifacts/0/docs/build/html/index.html) when [this CircleCI job](https://circleci.com/gh/mlflow/mlflow/37393) completes successfully.

<details>
<summary>More info</summary>

- Ignore this comment if this PR does not change the documentation.
- It takes a few minutes for the preview to be available.
- The preview is updated when a new commit is pushed to this PR.
- This comment was created by https://github.com/mlflow/mlflow/actions/runs/3627905851.

</details>

mshtelma committed:
Merge branch 'mlflow:master' into multiclassclassification

mshtelma committed:
Support Multi Class Classification

Signed-off-by: mshtelma <mshtelma@gmail.com>

mshtelma committed:
Support Multi Class Classification

Signed-off-by: mshtelma <mshtelma@gmail.com>

mshtelma committed:
Support Multi Class Classification

Signed-off-by: mshtelma <mshtelma@gmail.com>

mshtelma committed:
Support Multi Class Classification: Improved code

Signed-off-by: mshtelma <mshtelma@gmail.com>

mshtelma committed:
Support Multi Class Classification: Improved code

Signed-off-by: mshtelma <mshtelma@gmail.com>

sunishsheth2009 reviewed (commented):
I think the main question is about the design decision to reuse classification/v1 as the template name. @mengxr @jinzhang21 What do you guys think about this?

mengxr commented on mlflow/recipes/utils/metrics.py:
```diff
@@ -37,7 +37,7 @@ def from_custom_metric_dict(cls, custom_metric_dict):
 
-BUILTIN_CLASSIFICATION_RECIPE_METRICS = [
+BUILTIN_BINARY_CLASSIFICATION_RECIPE_METRICS = [
     RecipeMetric(name="true_negatives", greater_is_better=True),
```
Those metrics need to know which class is positive, which should be specified by the user.

mshtelma replied:
I have removed them from the multi-class dictionary.

mengxr commented on mlflow/recipes/utils/metrics.py:
```diff
@@ -46,6 +46,20 @@ def from_custom_metric_dict(cls, custom_metric_dict):
+    RecipeMetric(name="true_negatives", greater_is_better=True),
+    RecipeMetric(name="false_positives", greater_is_better=False),
+    RecipeMetric(name="false_negatives", greater_is_better=False),
+    RecipeMetric(name="true_positives", greater_is_better=True),
```
remove those metrics, which are for per-class not the aggregated metric

mshtelma replied:
Done!

mshtelma commented on mlflow/recipes/steps/evaluate.py:
```diff
@@ -61,19 +63,16 @@ def _validate_and_apply_step_config(self):
                 "Missing recipe config in recipe config.",
                 error_code=INVALID_PARAMETER_VALUE,
             )
-        if "positive_class" not in self.step_config and self.recipe == "classification/v1":
-            raise MlflowException(
-                "`positive_class` must be specified for classification/v1 recipes.",
-                error_code=INVALID_PARAMETER_VALUE,
-            )
```
Yes, it's not an error anymore. If the user does not define the positive_class parameter, the problem will be treated as a multi-class problem. We can add another parameter to let the user explicitly decide if he is dealing with a binary or multi-class classification problem instead of using the presence of a positive_class parameter to decide whether it's a binary or multi-class.  Does it make sense? @mengxr Is it ok to use positive_class for this decision?

jinzhang21 replied:
Even in a multi-class scenario, I can still define a "positive" class. Given the fact that we can detect whether it's a binary class or multi-class problem easily, I suggest to keep this warning for binary class problem.

mshtelma replied:
Ok, that makes sense. I have returned the warning for binary classification.

mshtelma commented on mlflow/recipes/steps/split.py:
```diff
@@ -26,7 +29,7 @@
 
 def _make_elem_hashable(elem):
-    import numpy as np
+    import numpy as np  # pylint: disable=reimported
```
ok! will remove the whole import. is it possible that it's used for the distributed hyperopt?

jinzhang21 commented on mlflow/recipes/utils/metrics.py:
```diff
@@ -86,6 +100,26 @@ def logloss(true_label, predicted_proba, eps=1e-15):
     )
 
 
+def _get_extended_task(step_config: Dict[str, Any]) -> str:
```
+1. All is needed here is `step_config["recipe"]` and `step_config["positive_class"]`

mshtelma replied:
Done.

jinzhang21 commented on mlflow/recipes/utils/metrics.py:
```diff
@@ -86,6 +100,26 @@ def logloss(true_label, predicted_proba, eps=1e-15):
+        return "regression"
+    elif "classification" in tmpl:
+        if "positive_class" in step_config:
+            return "binary_classification"
+        else:
+            return "multiclass_classification"
```
Can we call these `classification/binary` and `classification/multiclass` instead? Ultimately these should be very similar to the `recipe` field. I can imagine we might offer sub-recipes like `classification/binary/v1` and `classification/multiclass/v2`

mshtelma replied:
Great idea! Done.

mshtelma committed:
Merge branch 'mlflow:master' into multiclassclassification

mshtelma committed:
Merge branch 'mlflow:master' into multiclassclassification

mshtelma committed:
1) Class-aware splitting for multi-class use cases
2) Improved tests & and fixed several issues around metrics mapping for FLAML integration

Signed-off-by: Michael Shtelma <mshtelma@gmail.com>

mshtelma committed:
Added check to ensure that we have a positive class for a binary classification task

Signed-off-by: Michael Shtelma <mshtelma@gmail.com>

mshtelma committed:
Added check to ensure that we have a positive class for a binary classification task

Signed-off-by: Michael Shtelma <mshtelma@gmail.com>

mshtelma commented on mlflow/recipes/steps/ingest/__init__.py:
```diff
@@ -57,14 +57,15 @@ def _validate_and_apply_step_config(self):
+            #     raise MlflowException(
+            #         "`positive_class` must be specified for classification/v1 recipes.",
+            #         error_code=INVALID_PARAMETER_VALUE,
+            #     )
```
The warning is back, but the check is executed only for binary problems.

mshtelma committed:
Improved metric handling

Signed-off-by: Michael Shtelma <mshtelma@gmail.com>

mshtelma committed:
Merge branch 'mlflow:master' into multiclassclassification

jinzhang21 commented on mlflow/recipes/steps/split.py:
```diff
@@ -202,29 +240,46 @@ def _set_target_col_as_first(df, target_col):
+                    )
+                    class_weights = list(zip(classes, class_weights))
+                    class_weights = sorted(class_weights, key=lambda x: x[1], reverse=True)
+                    if len(class_weights) > 5:
```
Can we add a parameter `_MAX_CLASSES_TO_PROFILE = 5` to the top of this file?

mshtelma replied:
Done.

jinzhang21 commented on mlflow/recipes/steps/split.py:
```diff
@@ -202,29 +240,46 @@ def _set_target_col_as_first(df, target_col):
+                sub_title = (
+                    "Positive vs Negative"
+                    if self.positive_class is not None
+                    else "Different classes"
```
`Different classes` -> f`Top {min(5, len(class_weights))} Classes`

mshtelma replied:
Done.

mshtelma committed:
1)Added log_loss metrics
2)Refactored stratified split impl - improved performance almost 4x

Signed-off-by: Michael Shtelma <michael.shtelma@databricks.com>

mshtelma commented on mlflow/recipes/steps/automl/flaml.py:
```diff
@@ -15,15 +15,20 @@
     "r2_score": "r2",
     "mean_absolute_percentage_error": "mape",
     "f1_score": "f1",
+    "f1_score_micro": "micro_f1",
+    "f1_score_macro": "macro_f1",
     "accuracy_score": "accuracy",
+    "roc_auc": "roc_auc",
+    "roc_auc_ovr": "roc_auc_ovr",
+    "roc_auc_ovo": "roc_auc_ovo",
```
Done.

mshtelma commented on mlflow/recipes/steps/split.py:
```diff
@@ -38,6 +42,47 @@ def _make_elem_hashable(elem):
 
+def _run_split(task, input_df, split_ratios, target_col):
+    if task == "classification":
+        return _perform_split_per_class(input_df, split_ratios, target_col)
```
Using the current split implementation for multi-class problems we might run into the situation that the test or validation dataset will not get any examples of some heavily under-represented classes. This primarily affects multi-class classification but can also affect binary if we consider anomaly detection or fraud use cases where the positive (fraud) class is usually heavily under-represented. Our rebalancing feature cannot help here since we apply it to the train dataset only. As a result, users can get errors during the evaluation step because mlflow.evaluate() checks if all classes are represented in the test/validation dataset.

mshtelma replied:
I have changed the function name accordingly.

mshtelma replied:
>Stratification is a nice feature to have but is it always desired?
I do not see any drawbacks to classification use cases..

mshtelma commented on mlflow/recipes/steps/split.py:
```diff
@@ -38,6 +42,47 @@ def _make_elem_hashable(elem):
+        with ThreadPool(os.cpu_count() or _NUM_DEFAULT_CPUS) as p:
+            zipped_dfs = p.map(partial_func, classes)
+            test_df, train_df, validation_df = [pd.concat(x) for x in list(zip(*zipped_dfs))]
+            return test_df, train_df, validation_df
```
Yes, I know, but we parallelize only the hashing, not the whole process. 

The current implementation (in PR) runs all classes in parallel (pool with <CPU count> executors) and, after that, runs hashing also in parallel but with a reduced parallelization factor (Pool with two executors). For regression, it runs exactly as it was running before (Pool with eight threads). I know there are many ways to improve it even further, but I think the current approach is already not that bad. 

Using this approach, I have done the following measurements:
I have used a dataset containing 1134634 rows (616MB in parquet) and 17 classes. 
My new implementation needs 12 seconds for a stratified split. The old implementation (running one class after another) needs 46 seconds. I think it's a significant lift.

jinzhang21 replied:
Beautiful!

mshtelma commented on mlflow/recipes/steps/split.py:
```diff
@@ -38,6 +42,47 @@ def _make_elem_hashable(elem):
+
+def _perform_split_per_class(input_df, split_ratios, target_col):
+    classes = np.unique(input_df[target_col])
+    if len(classes) > 2:
```
Good point. Now I am using a stratified split for binary problems as well.

jinzhang21 reviewed (approved):
Very nice change, @mshtelma !

mshtelma committed:
Using X and y in transformer.fit call.

mshtelma committed:
Merge branch 'mlflow:master' into multiclassclassification

sunishsheth2009 reviewed (approved):
LGTM to me as well. Thank you :) 
Awesome work. I am sure the customers are eagerly waiting for it
